go-packages/meep-model/model.go

```diff
@@ -943,6 +957,737 @@ func (m *Model) GetNodeContext(name string) (ctx interface{}) {
+				nList = append(nList, nListOther...)
+			}
+		}
+	default:
+	}
+	return nList
+}
+
+// GetNodesByParentLevelFilter - Get list of nodes that match the filter values for all parents above the current hierarchy level provided
+//              Returned value is of type []*Node
+func (m *Model) GetNodesByParentLevelFilter(level string, filter *NodeFindFilter) []*Node {
+
+	var nodeList []*Node
+	nList := m.GetNodesQueriedLevelByFilter(level, filter)
+	//return immediately if no results
+	if nList == nil {
+		return nList
+	}
+	//levelsHierarchy := [5]string{"domain", "zone", "networkLocation", "physicalLocation", "process"}
+	//using an index value for a hierarchy level matching above string array
+	var initialLevelIndex int
+	//find current index level
+	switch level {
+	case "domain":
+		//no parents so return right away
+		//equivalent to currentLevelIndex = 0
+		return nList
+	case "zone":
+		initialLevelIndex = 1
+	case "networkLocation":
+		initialLevelIndex = 2
+	case "physicalLocation":
+		initialLevelIndex = 3
+	case "process":
+		initialLevelIndex = 4
+	default:
+		return nil
+	}
+
+	var parentName, parentType string
+	var parentFilterName, parentFilterType string
+
+	for _, currentNode := range nList {
+		addNode := true
+		parentNode := currentNode.parent
+		currentLevelIndex := initialLevelIndex
+		for currentLevelIndex != 0 {
+			switch currentLevelIndex {
+			case 1:
+				parentFilterName = filter.DomainName
+				parentFilterType = filter.DomainType
+				parentName = parentNode.(*dataModel.Domain).Name
+				parentType = parentNode.(*dataModel.Domain).Type_
+			case 2:
+				parentFilterName = filter.ZoneName
+				parentFilterType = ""
+				parentName = parentNode.(*dataModel.Zone).Name
+				parentType = parentNode.(*dataModel.Zone).Type_
+			case 3:
+				parentFilterName = filter.NetworkLocationName
+				parentFilterType = filter.NetworkLocationType
+				parentName = parentNode.(*dataModel.NetworkLocation).Name
+				parentType = parentNode.(*dataModel.NetworkLocation).Type_
+			case 4:
+				parentFilterName = filter.PhysicalLocationName
+				parentFilterType = filter.PhysicalLocationType
+				parentName = parentNode.(*dataModel.PhysicalLocation).Name
+				parentType = parentNode.(*dataModel.PhysicalLocation).Type_
+			default:
+				break
+			}
+
+			if parentFilterName != "" {
+				if parentName != parentFilterName {
+					addNode = false
+					break
+				}
+			}
+			if parentFilterType != "" {
+				if parentType != parentFilterType {
+					addNode = false
+					break
+				}
+			}
+			currentLevelIndex--
+			parentNode = m.GetNodeParent(parentName)
+		}
+		if addNode {
+			nodeList = append(nodeList, currentNode)
+		}
+	}
+	return nodeList
+}
+
+// GetDomainNodesByFilter - Get list of interfaces that match the filter provided for domains
+//              Returned value is of type []interface{}
+func (m *Model) GetDomainNodesByFilter(filter *NodeFindFilter) (node []interface{}) {
```

Comment:
These should return specific dataModel types instead of interface{}. The reason for creating several functions here is to perform type-specific processing.

Reply:
Each function should do the following:
1) Get a list of nodes for the required type
2) Pass each node through the filter function:
     - Filter function takes element type & filter criteria as input
3) Copy the filtered node list to a type-specific structure
     - Exclude children if necessary
     - Minimize if necessary
     - Structure is ready to be marshalled by the caller

Reply:
was restructured...